ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+    raise Temporal::Error::ActivityCancelled, 'cancellation details here'
+  end
+end
+```
+
+Notes:
+
+- `activity` is an instance of `Temporal::Activity::Context`
+- `activity.logger` will use a logging adapter configure via `Temporal::Configuration` (more on
+  this in the 2nd phase)
+- Common logging context can be configured by using middleware (more on this in the 2nd phase)
+- `activity.heartbeat` is a synchronous call
+- In case of a cancellation request `Temporal::Error::CancellationRequested` will get raised, which
+  can be accepted by raising `Temporal::Error::ActivityCancelled`
+
+```ruby
+class Temporal::Activity::Context
+  # Access to a pre-configured logger
+  def logger() -> Temporal::Logger
+
+  # Heartbeat will raise if a cancellation was requested
+  def heartbeat(*details) -> void
+
+  # Provides access to extra details from PollActivityTaskQueueResponse
+  def info() -> Temporal::Info::Activity
+end
+```
+
+### Return values and errors
+
+An activity can communicate its result to the workflow by returning a value or raising an error. The
+return values will get serialized as a `Payload` object similar to activity input arguments.
+Similarly, an error will get serialized and re-raised from within the workflow (if an activity
+called "synchronously").
+
+The main idea here is to make it feel (from the workflow's perspective) like a local invocation of
+an activity without any additional constructs in the way.
+
+```ruby
+class MakePaymentActivity < Temporal::Activity
+  PaymentMethodNotFound = Class.new(Temporal::Error::ActivityFailure)
+  PaymentFailed = Class.new(Temporal::Error::ActivityFailure)
+
+  def execute(payment_method_id)
+    payment_method = PaymentMethod.find(payment_method_id)
+    raise UserNotFound, "No payment_method with id #{payment_method_id}" unless payment_method
+
+    result = PaymentService.execute(user.payment_method)
+    raise PaymentFailed, "Unable to process payment: #{result.error}" if result.failure?
+
+    result.payment_id
+  end
+end
+```
+
+Notes:
+
+- Errors defined within the activity class are namespaced as `MyActivity::PaymentFailed`,
+  which is a very useful property
+  - This will however not work if a workflow does not have access to the activity definition
+  - In this case it will get raised as a generic `Temporal::Error::ActivityFailure` with an original
+    error's class name available via `#original_error`
+- Subclassing `Temporal::Error::ActivityFailure` allows us to inject extra information into the
+  error (such as activity name, id, etc)
+- All errors will also have their backtrace serialized and relayed to the caller
+
+
+## Payload converters
+
+Ruby SDK will use the model proposed [here](https://docs.temporal.io/docs/concepts/what-is-a-data-converter/)
+to facilitate conversion between SDK inputs and payloads. Basically there will be two extension
+points on the `Temporal::Configuration` object allowing to add either a DataConverter or a Codec.
```

Comment:
I wonder if we want to change the naming a bit here. In my mind, a "DataConverter" = "PayloadConverters" + "PayloadCodecs". But I understand then that the "DataConverter" has very little meaning since it's just a container and you can remove that abstraction. So maybe the names can be PayloadConverter and PayloadCodec? We generally tend to refer to "data conversion" as the combination of the two, not just one of the two. Would like other team member's input here.

Reply:
Sounds good to me, I've picked those names from — https://docs.temporal.io/docs/concepts/what-is-a-data-converter/, which you may wanna consider updating

Reply:
I think that document is accurate, it is just missing that the default data converter is a collection of payload converters.